realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -413,38 +414,47 @@ private static void initializeRealm(Realm realm) {
         // the Realm is initialized.
         boolean commitChanges = false;
         try {
-            realm.beginTransaction();
+            // We need to start a transaction no matter readOnly mode, because it acts as an interprocess lock.
+            // TODO: For proper inter-process support we also need to move e.g copying the asset file under an
+            // interprocess lock. This lock can obviously not be created by a Realm instance so we probably need
+            // to implement it in Object Store. When this happens, the `beginTransaction(true)` can be removed again.
+            realm.beginTransaction(true);
             long currentVersion = realm.getVersion();
             boolean unversioned = currentVersion == UNVERSIONED;
             commitChanges = unversioned;
 
             RealmConfiguration configuration = realm.getConfiguration();
 
-            if (unversioned) {
-                realm.setVersion(configuration.getSchemaVersion());
-            }
+            // Only allow creating the schema if not in read-only mode
```

Comment:
How about throwing some exception if `unversioned && configuration.isReadOnly()` is `true`?
I think it makes following code much simpler.